core/trino-main/src/main/java/io/trino/operator/WorkProcessorUtils.java

```diff
@@ -182,6 +183,24 @@ else if (processor.isBlocked()) {
         }
     }
 
+    static <T> WorkProcessor<T> blocking(WorkProcessor<T> processor, Supplier<ListenableFuture<Void>> futureSupplier)
+    {
+        requireNonNull(processor, "processor is null");
+        requireNonNull(futureSupplier, "futureSupplier is null");
+        return processor.transform(element -> {
+            if (element == null) {
+                return TransformationState.finished();
+            }
+
+            ListenableFuture<Void> future = futureSupplier.get();
+            if (!future.isDone()) {
+                return TransformationState.blocked(future);
```

Comment:
I presume this is safe due to how we use it in other places, but what happens if the future transitions to "done" between the check and the return?